File: src/main/java/frc/team449/robot2022/RobotContainer2022.java
package frc.team449.robot2022;

import edu.wpi.first.math.controller.PIDController;
import edu.wpi.first.math.controller.ProfiledPIDController;
import edu.wpi.first.math.controller.SimpleMotorFeedforward;
import edu.wpi.first.math.trajectory.TrapezoidProfile;
import edu.wpi.first.wpilibj.Encoder;
import edu.wpi.first.wpilibj.SerialPort;
import edu.wpi.first.wpilibj.XboxController;
import edu.wpi.first.wpilibj.smartdashboard.Field2d;
import edu.wpi.first.wpilibj.smartdashboard.SendableChooser;
import edu.wpi.first.wpilibj.smartdashboard.SmartDashboard;
import frc.team449.control.auto.AutoRoutine;
import frc.team449.control.holonomic.SwerveDrive;
import frc.team449.robot2022.drive.DriveConstants;
import frc.team449.system.AHRS;
import frc.team449.system.encoder.AbsoluteEncoder;
import frc.team449.system.encoder.BackupEncoder;
import frc.team449.system.encoder.NEOEncoder;
import frc.team449.system.encoder.QuadEncoder;
import frc.team449.system.motor.SparkMaxConfig;
import frc.team449.system.motor.WrappedMotor;
import io.github.oblarg.oblog.Loggable;

public final class RobotContainer2022 implements Loggable {

  // Other CAN IDs
  public static final int PDP_CAN = 1, PCM_MODULE = 0;

  public final XboxController driveJoystick = new XboxController(0);

  public final AHRS ahrs = new AHRS(SerialPort.Port.kMXP);

  public final SwerveDrive drive;
  // Instantiate/declare PDP and other stuff here

  public final Field2d field = new Field2d();
  public final SendableChooser<AutoRoutine> autoChooser = new SendableChooser<>();

  public RobotContainer2022() {
    SmartDashboard.putData(field);
    this.drive = createDrivetrain();
  }

  /** Helper to make turning motors for swerve */
  private static WrappedMotor makeDrivingMotor(
    String name,
    int motorId,
    boolean inverted,
    Encoder wpiEnc
  ) {
    return new SparkMaxConfig()
      .setName(name + "Drive")
      .setId(motorId)
      .setEnableBrakeMode(true)
      .setInverted(inverted)
      .setEncoderCreator(
        BackupEncoder.creator(
          QuadEncoder.creator(
            wpiEnc,
            DriveConstants.DRIVE_EXT_ENC_CPR,
            DriveConstants.DRIVE_UPR,
            1
          ),
          NEOEncoder.creator(
            DriveConstants.DRIVE_UPR,
            DriveConstants.DRIVE_GEARING
          ),
          DriveConstants.DRIVE_ENC_VEL_THRESHOLD
        )
      )
      .build();
  }

  /** Helper to make turning motors for swerve */
  private static WrappedMotor makeTurningMotor(
    String name,
    int motorId,
    boolean inverted,
    int encoderChannel,
    double offset
  ) {
    return new SparkMaxConfig()
      .setName(name + "Turn")
      .setId(motorId)
      .setEnableBrakeMode(true)
      .setInverted(inverted)
      .setEncoderCreator(
        AbsoluteEncoder.creator(
          encoderChannel,
          2 * Math.PI,
          offset,
          DriveConstants.TURN_UPR,
          DriveConstants.TURN_GEARING
        )
      )
      .build();
  }

  private SwerveDrive createDrivetrain() {
    // todo actually make the modules
    return SwerveDrive.squareDrive(
      ahrs,
      DriveConstants.MAX_LINEAR_SPEED,
      DriveConstants.MAX_ROT_SPEED,
      makeDrivingMotor(
        "FL",
        DriveConstants.FRONT_LEFT_DRIVE,
        false,
        DriveConstants.DRIVE_ENC_FL
      ),
      makeDrivingMotor(
        "FR",
        DriveConstants.FRONT_RIGHT_DRIVE,
        false,
        DriveConstants.DRIVE_ENC_FR
      ),
      makeDrivingMotor(
        "BL",
        DriveConstants.BACK_LEFT_DRIVE,
        false,
        DriveConstants.DRIVE_ENC_BL
      ),
      makeDrivingMotor(
        "BR",
        DriveConstants.BACK_RIGHT_DRIVE,
        false,
        DriveConstants.DRIVE_ENC_BR
      ),
      makeTurningMotor(
        "FL",
        DriveConstants.FRONT_LEFT_TURN,
        false,
        DriveConstants.TURN_ENC_CHAN_FL,
        DriveConstants.TURN_ENC_OFFSET_FL
      ),
      makeTurningMotor(
        "FR",
        DriveConstants.FRONT_LEFT_TURN,
        false,
        DriveConstants.TURN_ENC_CHAN_FR,
        DriveConstants.TURN_ENC_OFFSET_FR
      ),
      makeTurningMotor(
        "BL",
        DriveConstants.FRONT_LEFT_TURN,
        false,
        DriveConstants.TURN_ENC_CHAN_BL,
        DriveConstants.TURN_ENC_OFFSET_BL
      ),
      makeTurningMotor(
        "BR",
        DriveConstants.FRONT_LEFT_TURN,
        false,
        DriveConstants.TURN_ENC_CHAN_BR,
        DriveConstants.TURN_ENC_OFFSET_BR
      ),
      DriveConstants.FRONT_LEFT_LOC,
      () -> new PIDController(0, 0, 0),
      () ->
        new ProfiledPIDController(
          0,
          0,
          0,
          new TrapezoidProfile.Constraints(0, 0)
        ),
      new SimpleMotorFeedforward(0, 0, 0),
      new SimpleMotorFeedforward(0, 0, 0)
    );
  }

  public void teleopInit() {
    // todo Add button bindings here
  }

  public void robotPeriodic() {
    // todo Update robot pose on field and stuff
  }

  public void simulationPeriodic() {
    // Update simulated mechanisms on Mechanism2d widget and stuff
  }
}

File: src/main/kotlin/frc/team449/robot2022/RobotContainer2022.kt
package frc.team449.robot2022

import edu.wpi.first.math.controller.PIDController
import edu.wpi.first.math.controller.SimpleMotorFeedforward
import edu.wpi.first.math.filter.SlewRateLimiter
import edu.wpi.first.wpilibj.PowerDistribution
import edu.wpi.first.wpilibj.RobotBase.isReal
import edu.wpi.first.wpilibj.SerialPort
import edu.wpi.first.wpilibj.XboxController
import edu.wpi.first.wpilibj.smartdashboard.SendableChooser
import frc.team449.RobotContainerBase
import frc.team449.control.auto.AutoRoutine
import frc.team449.control.holonomic.OIHolonomic
import frc.team449.control.holonomic.SwerveDrive
import frc.team449.robot2022.auto.Example
import frc.team449.robot2022.drive.DriveConstants
import frc.team449.system.AHRS
import frc.team449.system.encoder.AbsoluteEncoder
import frc.team449.system.encoder.NEOEncoder
import frc.team449.system.motor.createSparkMax
import io.github.oblarg.oblog.annotations.Log
import kotlin.math.abs

class RobotContainer2022() : RobotContainerBase() {

  // Other CAN IDs
  val PDP_CAN = 1
  val PCM_MODULE = 0

  private val driveController = XboxController(0)

  private val ahrs = AHRS(SerialPort.Port.kMXP)

  // Instantiate/declare PDP and other stuff here

  override val powerDistribution: PowerDistribution = PowerDistribution(PDP_CAN, PowerDistribution.ModuleType.kCTRE)

  /**
   * Converts the drive to a SwerveSim if the robot is in simulation
   */
  @Log.Include
  override val drive = if (isReal()) createDrivetrain() else SwerveDrive.simOf(createDrivetrain())

  override val autoChooser = addRoutines()

  override val oi = OIHolonomic(
    drive,
    { if (abs(driveController.leftY) < .08) .0 else driveController.leftY },
    { if (abs(driveController.leftX) < .08) .0 else driveController.leftX },
    { if (abs(driveController.getRawAxis(4)) < .02) .0 else driveController.getRawAxis(4) },
    SlewRateLimiter(1.5),
    2.5,
    true
  )

  /** Helper to make turning motors for swerve */
  private fun makeDrivingMotor(
    name: String,
    motorId: Int,
    inverted: Boolean
  ) =
    createSparkMax(
      name = name + "Drive",
      id = motorId,
      enableBrakeMode = true,
      inverted = inverted,
      encCreator =
      NEOEncoder.creator(
        DriveConstants.DRIVE_UPR,
        DriveConstants.DRIVE_GEARING
      )
    )

  /** Helper to make turning motors for swerve */
  private fun makeTurningMotor(
    name: String,
    motorId: Int,
    inverted: Boolean,
    sensorPhase: Boolean,
    encoderChannel: Int,
    offset: Double
  ) =
    createSparkMax(
      name = name + "Turn",
      id = motorId,
      enableBrakeMode = true,
      inverted = inverted,
      encCreator = AbsoluteEncoder.creator(
        encoderChannel,
        offset,
        DriveConstants.TURN_UPR,
        sensorPhase
      )
    )

  private fun createDrivetrain() =
    SwerveDrive.squareDrive(
      ahrs,
      DriveConstants.MAX_LINEAR_SPEED,
      DriveConstants.MAX_ROT_SPEED,
      makeDrivingMotor(
        "FL",
        DriveConstants.DRIVE_MOTOR_FL,
        false
      ),
      makeDrivingMotor(
        "FR",
        DriveConstants.DRIVE_MOTOR_FR,
        false
      ),
      makeDrivingMotor(
        "BL",
        DriveConstants.DRIVE_MOTOR_BL,
        false
      ),
      makeDrivingMotor(
        "BR",
        DriveConstants.DRIVE_MOTOR_BR,
        false
      ),
      makeTurningMotor(
        "FL",
        DriveConstants.TURN_MOTOR_FL,
        true,
        false,
        DriveConstants.TURN_ENC_CHAN_FL,
        DriveConstants.TURN_ENC_OFFSET_FL
      ),
      makeTurningMotor(
        "FR",
        DriveConstants.TURN_MOTOR_FR,
        true,
        false,
        DriveConstants.TURN_ENC_CHAN_FR,
        DriveConstants.TURN_ENC_OFFSET_FR
      ),
      makeTurningMotor(
        "BL",
        DriveConstants.TURN_MOTOR_BL,
        true,
        false,
        DriveConstants.TURN_ENC_CHAN_BL,
        DriveConstants.TURN_ENC_OFFSET_BL
      ),
      makeTurningMotor(
        "BR",
        DriveConstants.TURN_MOTOR_BR,
        true,
        false,
        DriveConstants.TURN_ENC_CHAN_BR,
        DriveConstants.TURN_ENC_OFFSET_BR
      ),
      DriveConstants.FRONT_LEFT_LOC,
      { PIDController(DriveConstants.DRIVE_KP, DriveConstants.DRIVE_KI, DriveConstants.DRIVE_KD) },
      {
        PIDController(DriveConstants.TURN_KP, DriveConstants.TURN_KI, DriveConstants.TURN_KD)
      },
      SimpleMotorFeedforward(DriveConstants.DRIVE_KS, DriveConstants.DRIVE_KV, DriveConstants.DRIVE_KA),
      SimpleMotorFeedforward(DriveConstants.TURN_KS, DriveConstants.TURN_KV, DriveConstants.TURN_KA)
    )

  private fun addRoutines(): SendableChooser<AutoRoutine> {
    val chooser = SendableChooser<AutoRoutine>()
    val exampleAuto = Example("Example", 2.0, 2.0, drive)
    chooser.setDefaultOption("Example Swerve Auto", exampleAuto.routine())

    return chooser
  }

  override fun teleopInit() {
    // todo Add button bindings here
  }

  override fun robotPeriodic() {
  }

  override fun simulationInit() {
    // DriverStationSim.setEnabled(true)
  }

  override fun simulationPeriodic() {
    // Update simulated mechanisms on Mechanism2d widget and stuff
  }
}
